components/ai_chat/core/browser/engine/oai_parsing.cc

```diff
@@ -0,0 +1,168 @@
+
+#include "brave/components/ai_chat/core/browser/engine/oai_parsing.h"
+
+#include <string>
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "brave/components/ai_chat/core/browser/engine/engine_consumer.h"
+#include "brave/components/ai_chat/core/browser/tools/tool.h"
+
+namespace ai_chat {
+
+namespace {
+
+// Prevent parsing arguments list that's too large for the
+// tools we're providing.
+const size_t kMaxArgumentsLength = 1000000;
+
+// Prevent hallucinated JSON structures
+const size_t kMaxArgumentsJSONDepth = 20;
+
+}  // namespace
+
+std::vector<mojom::ToolUseEventPtr> ToolUseEventFromToolCallsResponse(
+    const base::Value::List* tool_calls_api_response) {
+  // https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools
+  // https://platform.openai.com/docs/api-reference/chat/object
+  // choices -> message -> tool_calls
+  std::vector<mojom::ToolUseEventPtr> tool_use_events;
+  for (auto& tool_call_raw : *tool_calls_api_response) {
+    if (!tool_call_raw.is_dict()) {
+      DLOG(ERROR) << "Tool call is not a dictionary.";
+      continue;
+    }
+    const auto& tool_call = tool_call_raw.GetDict();
+    mojom::ToolUseEventPtr tool_use_event = mojom::ToolUseEvent::New();
+
+    const base::Value::Dict* function = tool_call.FindDict("function");
+    if (!function) {
+      DLOG(ERROR) << "No function info found in tool call.";
+      continue;
+    }
+    // Tool call results can be partial and should be added to the previous
+    // event by the event handler.
+    const std::string* id = tool_call.FindString("id");
+    if (!id) {
+      DLOG(ERROR) << "No id found in tool call.";
+      continue;
+    }
+    tool_use_event->id = *id;
+
+    const std::string* name = function->FindString("name");
+    if (!name) {
+      DLOG(ERROR) << "No name found in tool call.";
+      continue;
+    }
+    tool_use_event->tool_name = *name;
+
+    // Arguments are provided as a JSON string. We parse it here so that the
+    // rest of the code doesn't have to parse it potentially multiple times
+    // (performance), and we can add some protections.
+    const std::string* arguments_raw = function->FindString("arguments");
+    if (!arguments_raw) {
+      DLOG(ERROR) << "No arguments found in tool call.";
+      continue;
+    }
+
+    // We can be defensive about hallucinated large JSON and structures.
+    // However, we still want to allow for large content as valid tool use can
+    // be providing lengthy answers via tools in order to force the LLM to
+    // provide the answer with a specific JSON structure.
+    if (arguments_raw->size() > kMaxArgumentsLength) {
+      DLOG(ERROR) << "Tool call arguments are too large: "
+                  << arguments_raw->size() << " bytes.";
+      continue;
+    }
+
+    auto arguments =
+        base::JSONReader::ReadDict(*arguments_raw, /*options=*/{},
+                                   /*max_depth=*/kMaxArgumentsJSONDepth);
```

Comment:
This is currently only used by our conversation api client and response from our server is considered trusted since we do sanitization on server but if in the future we use this function on BYOM, shouldn't we sanitizing the input?

Reply:
ugh yeah you're probably right since `APIRequestHelper`'s JSON sanitizer will likely not cover this part since it's hidden in a string? I'm not sure to be honest what DataDecoder protects against and if this string will already be sanitized by `APIRequestHelper`'s previous parse.

It will be a pain to convert this to async but I'll take a look.